tee-worker/omni-executor/intent/executors/cross-chain/src/cross_chain.rs

```diff
@@ -97,20 +119,72 @@ impl<BinanceClient: BinanceApi, SolanaClient: SolanaClientTrait>
 		.await?;
 
 		// TODO: this should be abstracted away
-		let (payout_amount, payout_amount_u256) = self
-			.do_binance_swap(
-				omni_account,
-				swap_order.from_asset.clone(),
-				from_amount,
-				from_address,
-				pumpx_config.wallet_index,
-				false,
-			)
-			.await?;
+		// todo we only transfer to binance without swap
+
+		let should_wait_for_deposit_confirm = false;
+
+		self.do_binance_deposit(
+			omni_account,
+			swap_order.from_asset.clone(),
+			from_amount_decimal,
+			from_address.clone(),
+			pumpx_config.wallet_index,
+			should_wait_for_deposit_confirm,
+		)
+		.await?;
+
+		if instant {
+			// we should wait for deposit confirmation in separate thread to not block intent execution and release lock upon it
+			let amount_to_release = AmountType::from_str(&from_amount.to_string()).unwrap();
+			self.account_asset_lock.release(account_id.clone(), swap_order.from_asset.clone(), amount_to_lock
+				).map_err(|e| 	log::error!("Could not release asset lock for account: {:?}, amount: {:?}, asset: {:?}, reason: {:?}", account_id, swap_order.from_asset, amount_to_release, e))?;
+		}
 
-		self.do_payout(&payout_address, &payout_amount_u256).await?;
+		let mut payout_amount =
+			self.estimate_payout(swap_order.from_asset.clone(), from_amount_decimal).await?;
+		let mut payout_amount_u256 = str_to_u256(&payout_amount, 18)?;
+
+		// trade only if deposit is confirmed
+		if should_wait_for_deposit_confirm {
+			(payout_amount, payout_amount_u256) = self
+				.do_binance_trade(
+					swap_order.from_asset.clone(),
+					from_amount_decimal,
+					from_address.clone(),
+				)
+				.await?;
+		}
 
-		self.apply_gas_fee(&payout_amount, access_token, pumpx_config).await
+		// Fetch contract balance
+		let balance = self.accounting_contract_client.get_balance().await?;
+		if balance < payout_amount_u256 {
+			error!(
+				"There is not enough balance in the accounting contract, {} < {}",
+				balance, payout_amount_u256
+			);
+			return Err(());
+		}
+
+		if !instant {
+			let to_wallet = self
+				.pumpx_signer_client
+				.request_wallet(to_chain_type, pumpx_config.wallet_index, omni_account)
+				.await
+				.map_err(|e| error!("Could not get to_wallet from pumpx-signer: {:?}", e))?;
+
+			let to_address = pubkey_to_address(to_chain_type, &to_wallet)?;
+
+			let payout_address = Address::from_str(&to_address).map_err(|_| {
+				error!("Failed to parse payout address");
+			})?;
+			self.do_payout(&payout_address, &payout_amount_u256).await?;
```

Comment:
What does `instant` means? should we do the payout if `instant` is true?

Reply:
There is not payout for instant flow.
Accounting contract is used as source account for single chain swap.